tools/proc_init.py

```diff
@@ -0,0 +1,302 @@
+        node_features = recv_node_features_obj(rank, size)
+        edge_data = recvData( rank, 2, torch.int32 )
+
+    '''
+    sort data by type-id, and follow through with assigning globalIds
+    '''
+    type_ids = np.unique( node_data[ :, 0 ] )
+    type_ids.sort ()
+
+    lstNodeTypeCounts = []
+    typeCounts = np.bincount( node_data[ :, 0 ] )
+    for typeId in type_ids: 
+        lstNodeTypeCounts.append( (typeId, typeCounts[ typeId ]) )
+    print( 'Rank: ', rank, ', LstNodeTypeCounts: ', lstNodeTypeCounts )
+        
+    # after this call node_data = [ globalId, node_type, orig_node_type_id, line_id, local_type_id ]
+    node_data, rank_offset_globalid = assignGlobalNodeIds ( rank, size, lstNodeTypeCounts, node_data )
+    print( 'Rank: ', rank, ' Done assign Global ids to nodes...')
+
+    #Work on the edge and assign GlobalIds
+    etype_ids = np.unique( edge_data[:, 3] )
+    etype_ids.sort ()
+    print('Rank: ', rank, ' etype-ids: ', etype_ids )
+    
+    edge_data = edge_data[ edge_data[ :, 3 ].argsort() ]
+    lstEdgeTypeCounts = []
+    edgeCounts = np.bincount( edge_data[ :,3 ] )
+
+    for etype in etype_ids: 
+        lstEdgeTypeCounts.append( (etype, edgeCounts[etype]) )
+    print('Rank: ', rank, ', LstEdgeTypeCounts: ', lstEdgeTypeCounts )
+
+    edge_data = assignGlobalEdgeIds ( rank, size, lstEdgeTypeCounts, edge_data )
+    print( 'Rank: ', rank, ' Done assign Global ids to edges...')
+
+    edge_data = getGlobalIdsForEdges( rank, size, edge_data, metis_partitions, node_data)
+    print( 'Rank: ', rank, ' Done retrieving Global Node Ids for non-local nodes... ' )
+    print( 'Rank: ', rank, ' Done with distributed processing... Starting to serialize necessary files per Metis partitions' )
+
+    '''
+    Here use the functionality of the convert partition.py to store the dgl objects
+    for the node_features, edge_features and graph itself. 
+    Also generate the json file for the entire graph. 
+    '''
+    pipelineArgs = {}
+    pipelineArgs[ "rank" ] = rank
+    pipelineArgs["node-global-id" ] = node_data[: ,0]
+    pipelineArgs["node-ntype"] = node_data[: ,1]
+    pipelineArgs["node-ntype-orig-ids" ] = node_data[:, 2]
+    pipelineArgs["node-orig-id" ] = node_data[:, 3]
+    pipelineArgs["node-local-node-type-id" ] = node_data[:, 4]
+    pipelineArgs["node-global-node-id-offset"] = rank_offset_globalid
+    pipelineArgs["node-features"] = node_features
+
+    #pipelineArgs[ "edge_data" ] = edge_data
+    pipelineArgs[ "edge-src-id" ] = edge_data[:,0]
+    pipelineArgs[ "edge-dst-id" ] = edge_data[:,1]
+    pipelineArgs[ "edge-orig-src-id" ] = edge_data[:,2]
+    pipelineArgs[ "edge-orig-dst-id" ] = edge_data[:,3]
+    pipelineArgs[ "edge-orig-edge-id" ] = edge_data[:,4]
+    pipelineArgs[ "edge-etype-ids" ] = edge_data[:,5]
+
+    #call convert_parititio.py for serialization 
+    json_metadata, output_dir, graph_name = processMetisPartitions (False, pipelineArgs, params)
```

Comment:
it seems to me that you have to save all data to the files and then this function loads data from the files. i thought we agree that we'll pass the shuffled data to this function directly.